rust/foxglove/src/schemas_wkt.rs

```diff
@@ -0,0 +1,238 @@
+        (sec.into() + i64::from(div), i32::try_from(nsec).unwrap())
+    }
+}
+
+/// A signed, fixed-length span of time.
+///
+/// The duration is represented by a count of seconds (which may be negative), and a count of
+/// fractional seconds at nanosecond resolution (which are always positive).
+///
+/// # Example
+///
+/// ```
+/// use foxglove::schemas::Duration;
+///
+/// // A duration of 2.718... seconds.
+/// let duration = Duration {
+///     sec: 2,
+///     nsec: 718_281_828,
+/// };
+///
+/// // A duration of -3.14... seconds. Note that nanoseconds are always in the positive
+/// // direction.
+/// let duration = Duration {
+///     sec: -4,
+///     nsec: 858_407_346,
+/// };
+///
+/// // Positive durations can be derived from std::time::Duration.
+/// let duration = std::time::Duration::from_micros(577_215);
+///
+/// // When built with the `chrono` feature, durations can also be derived from chrono::TimeDelta.
+/// #[cfg(feature = "chrono")]
+/// let duration = chrono::TimeDelta::microseconds(1_414_213);
+/// ```
+#[derive(Debug, Default, Clone, Copy, PartialEq)]
+pub struct Duration {
+    /// Seconds offset.
+    pub sec: i32,
+    /// Nanoseconds offset in the positive direction.
+    pub nsec: u32,
+}
+
+impl Duration {
+    /// Maximum representable duration.
+    pub const MAX: Self = Self {
+        sec: i32::MAX,
+        nsec: u32::MAX,
+    };
+
+    /// Minimum representable duration.
+    pub const MIN: Self = Self {
+        sec: i32::MIN,
+        nsec: 0,
+    };
+
+    fn into_prost(self) -> prost_types::Duration {
+        self.into()
+    }
+}
+
+impl From<Duration> for prost_types::Duration {
+    fn from(v: Duration) -> Self {
+        let (seconds, nanos) = normalize(v.sec, v.nsec);
+        Self { seconds, nanos }
+    }
+}
+
+impl prost::Message for Duration {
+    fn encode_raw(&self, buf: &mut impl bytes::BufMut)
+    where
+        Self: Sized,
+    {
+        self.into_prost().encode_raw(buf);
+    }
+
+    fn merge_field(
+        &mut self,
+        _tag: u32,
+        _wire_type: prost::encoding::wire_type::WireType,
+        _buf: &mut impl bytes::Buf,
+        _ctx: prost::encoding::DecodeContext,
+    ) -> Result<(), prost::DecodeError>
+    where
+        Self: Sized,
+    {
+        // We only support encoding for now.
+        unimplemented!("not implemeneted");
+    }
+
+    fn encoded_len(&self) -> usize {
+        self.into_prost().encoded_len()
+    }
+
+    fn clear(&mut self) {
+        self.sec = 0;
+        self.nsec = 0;
+    }
+}
+
+impl TryFrom<std::time::Duration> for Duration {
+    type Error = FoxgloveError;
+
+    fn try_from(duration: std::time::Duration) -> Result<Self, Self::Error> {
+        let Ok(sec) = i32::try_from(duration.as_secs()) else {
+            return Err(FoxgloveError::DurationOutOfRange);
+        };
+        let nsec = duration.subsec_nanos();
+        Ok(Self { sec, nsec })
+    }
+}
+
+/// A timestamp, represented as an offset from the unix epoch.
```

Comment:
If you're not working with and converting from std/chrono types, I don't think we need to make this claim. MCAP is specifically agnostic about the choice of epoch.

But the limitations, comments, and fallible conversions make sense otherwise.